core/arch/arm/kernel/user_ta.c

```diff
@@ -167,14 +173,12 @@ static TEE_Result get_elf_segments(struct user_ta_elf *elf,
 
 	idx = 1;
 	while (idx < num_segs) {
-		size_t this_size = segs[idx].oend - segs[idx].offs;
-		size_t prev_size = segs[idx - 1].oend - segs[idx - 1].offs;
-
-		if (core_is_buffer_intersect(segs[idx].offs, this_size,
-					     segs[idx - 1].offs, prev_size)) {
+		if (core_is_buffer_intersect(segs[idx].va, segs[idx].size,
+					     segs[idx - 1].va,
+					     segs[idx - 1].size)) {
 			/* Merge the segments and their attributes */
-			segs[idx - 1].oend = MAX(segs[idx - 1].oend,
-						 segs[idx].oend);
+			segs[idx - 1].size = segs[idx].va + segs[idx].size -
+					     segs[idx - 1].va;
```

Comment:
This assumes `segs[idx]` vmem range ends after `segs[idx -1]`.
Code below assumes `segs[idx]` vmem range **starts** after `segs[idx -1]` range.
Maybe few `assert()` would ensure those.

Reply:
Agree, I'll fix.